### ISSUE TITLE
Unable to properly save field with multiple inputs when repeated

### ISSUE THREAD

**Issue by adamcapriola** [WebDevStudios/CMB2#116]
I've been working with the <a href="https://github.com/WebDevStudios/CMB2/wiki/Adding-your-own-field-types#example-4-multiple-inputs-one-field-lets-create-an-address-field">address field example</a> and it appears that it is incompatible with the repeating option. Any data entered doesn't save properly and a new row is generated on each attempted save.

Pre-save:
![screen shot 2014-12-12 at 1 51 19 pm](https://cloud.githubusercontent.com/assets/279928/5417318/0e83c760-8206-11e4-8d66-9e22ad498033.png)
---
Post-save:
![screen shot 2014-12-12 at 1 51 41 pm](https://cloud.githubusercontent.com/assets/279928/5417321/125660f0-8206-11e4-988e-30d16e4df8b8.png)
---
Code (along with the <a href="https://github.com/WebDevStudios/CMB2/wiki/Adding-your-own-field-types#example-4-multiple-inputs-one-field-lets-create-an-address-field">address example</a>):

```php
add_filter( 'cmb2_meta_boxes', 'ac_metaboxes' );

function ac_metaboxes( array $meta_boxes ) {

	$meta_boxes['my_metabox'] = array(
	    'id'            => 'my_metabox',
	    'title'         => 'My Title',
	    'object_types'  => array( 'post' ),
	    'context'       => 'normal',
	    'priority'      => 'high',
	    'show_names'    => true,
	    'fields'        => array(
			array(
			    'name'       => 'Address',
			    'id'         => 'address',
			    'type'       => 'address',
			    'repeatable' => true
			)
	    )
	);

}
```

WordPress version: 4.0.1
CMB2 version: 2.0.0.3

**Comment by jtsternberg** [comment]
I've added a full-example of the address field in the Snippets Library: [address-field-type.php](https://github.com/WebDevStudios/CMB2-Snippet-Library/blob/master/custom-field-types/address-field-type.php)

**Comment by olya-shamian** [comment]
To those who wants to get in on the address-field-type links to the Snippet Library, it is now here: https://github.com/WebDevStudios/CMB2-Snippet-Library/tree/master/custom-field-types/address-field-type

**Comment by dreiacht** [comment]
Hey everyone,

first of all thanks for this great solution! Unfortunately I'm running into the following issue:
When adding 3 or more addresses at once, the last one will not be saved. So when I add 3 addresses only 2 will save, when adding 4 only 3 will save, etc. 

Any idea why that could be?

**Comment by tw2113** [comment]
@dreiacht if you could provide your configuration code, we can look into it and try to see what may be going on.

**Comment by dreiacht** [comment]
@tw2113 Thanks!

So, everything is in my cmb2.php file.

At first I include the address snippet / plugin.

`function cmb2_init_address_field() {
	require_once dirname( __FILE__ ) . '/cmb2/plugins/class-cmb2-render-address-field.php';
	CMB2_Render_Address_Field::init();
}
add_action( 'cmb2_init', 'cmb2_init_address_field' );`

And then I have a group field in which I place the repeatable address field:

`$guideline_group_field = $cmb_guidelines->add_field( array(
			'id'          => $prefix . 'guideline_sections',
			'type'        => 'group',
			'options'     => array(
				'group_title'   => esc_html__( 'Section {#}', 'bonestheme' ), // {#} gets replaced by row number
				'add_button'    => esc_html__( 'Add Another Section', 'bonestheme' ),
				'remove_button' => esc_html__( 'Remove Section', 'bonestheme' ),
				'sortable'      => true, // beta
				// 'closed'     => true, // true to have the groups closed by default
			),
		) );

$cmb_guidelines->add_group_field( $guideline_group_field, array(
				'name'       		=> esc_html__( 'Address', 'bonestheme' ),
				'id'         		=> 'cmb_address',
				'classes' 			=> 'cmb-section-address cmb-subsection',
				'type' 					=> 'address',
				'repeatable' 		=> 'true',
				'text' 					=> array(
					'add_row_text' 		=> 'Add Address',
					'remove_row_text' => 'Remove Address',
				),
			) );`

**Comment by tw2113** [comment]
Can you link to the repo that provides the contents of class-cmb2-render-address-field.php or simply provide that file itself, so we can reproduce as close as possible.

**Comment by dreiacht** [comment]
Here you go: https://github.com/WebDevStudios/CMB2-Snippet-Library/blob/master/custom-field-types/address-field-type/class-cmb2-render-address-field.php

**Comment by tw2113** [comment]
Ah, fair enough, it's from the snippet library :D

**Comment by tw2113** [comment]
So I put a breakpoint on the sanitize method in the address field class, and filled in 3 different addresses in one group. I then hit submit and waited for the breakpoint to catch. The meta values I saw come in was an array of 3 indexes. In each index was the keys specified, address 1, address 2, etc. However the 3rd index had an array empty of values. `'address_1' => ''` and so on. Towards the end, the plugin does an array_filter, which is going to clear out any resulting empty items. Thus why we're losing items past a certain amount.

The question turns to why is there an empty version being submitted? I moved my focus to the repeater and how you can dynamically fill in new groups for a new address.

First one, for an abbreviated example, has ID attributes of  `_school_guideline_sections_0_cmb_address_address_1_0` and `_school_guideline_sections_0_cmb_address_address_2_0` for the first grouping. Add a new one, and we get `_school_guideline_sections_0_cmb_address_address_1_1` and `_school_guideline_sections_0_cmb_address_address_2_1` First number at the end is for which address we're dealing with, and then the last part is which index in the metabox.

Here's where it gets interesting, on the 3rd grouping, we get `_school_guideline_sections_0_cmb_address_address_2_1` and then `_school_guideline_sections_0_cmb_address_address_2_2`

On the forth, I'm seeing `_school_guideline_sections_0_cmb_address_address_2_2` and `_school_guideline_sections_0_cmb_address_address_2_2`

For whatever reason, the incrementer/iterator for the repeatable aspect is getting confused and not assigning attributes properly. Looks like it's doing it for the rest of them as well. the city field is getting stuck at 2 for example.

Soooooo, my hunch is something's up with the javascript in this case, and how it's handling repeating the fields in this custom field type.

**Comment by valentin-debris** [comment]
I investigated this error and I found a solution.

The problem is in cmb2, when you click on the "Add row", the js function (in cmb2.js) "cmb.addAjaxRow" retrieves the data in the attribute "data-iterator" to determine the next value. This attribute is add to the label and the input. The prob is the fonction that retrieves the "data-iterator" always takes the first elem, the one in the label. Or the label stop incrementing after the second rows, while the data-interator on the input continues.

So they are 2 solutions :
-Remove all the labels in the class "class-cmb2-render-address-field"
-in cmb2.js (~ line 562, and cmb2.min.js), just add ".last()" in "cmb.addAjaxRow" function when calculating prevNum, between the find() and data() :
```
cmb.addAjaxRow = function( evt ) {
	evt.preventDefault();

	var $this         = $( this );
	var $table        = $id( $this.data('selector') );
	var $emptyrow     = $table.find('.empty-row');
	var prevNum       = parseInt( $emptyrow.find('[data-iterator]').last().data('iterator'), 10 );
...
}
```

ps : I test it
